src/main/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordData.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.bson.BsonDocument;
+
+import com.mongodb.MongoNamespace;
+import com.mongodb.client.model.WriteModel;
+
+import com.mongodb.kafka.connect.sink.converter.SinkConverter;
+import com.mongodb.kafka.connect.sink.converter.SinkDocument;
+import com.mongodb.kafka.connect.sink.writemodel.strategy.WriteModelStrategyHelper;
+
+final class MongoProcessedSinkRecordData {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoProcessedSinkRecordData.class);
+  private static final SinkConverter SINK_CONVERTER = new SinkConverter();
+
+  private final MongoSinkTopicConfig config;
+  private final MongoNamespace namespace;
+  private final SinkRecord sinkRecord;
+  private final SinkDocument sinkDocument;
+  private final WriteModel<BsonDocument> writeModel;
+
+  MongoProcessedSinkRecordData(final SinkRecord sinkRecord, final MongoSinkConfig sinkConfig) {
+    this.sinkRecord = sinkRecord;
+    this.config = sinkConfig.getMongoSinkTopicConfig(sinkRecord.topic());
+    this.sinkDocument = SINK_CONVERTER.convert(sinkRecord);
+    this.namespace = createNamespace();
+    this.writeModel = createWriteModel();
+  }
+
+  public MongoSinkTopicConfig getConfig() {
+    return config;
+  }
+
+  public MongoNamespace getNamespace() {
+    return namespace;
+  }
+
+  public SinkRecord getSinkRecord() {
+    return sinkRecord;
+  }
+
+  public WriteModel<BsonDocument> getWriteModel() {
+    if (writeModel == null) {
+      throw new DataException("Unable to create a valid WriteModel for the SinkRecord");
+    }
+    return writeModel;
+  }
+
+  public boolean canProcess() {
+    return namespace != null && writeModel != null;
+  }
+
+  MongoNamespace createNamespace() {
+    return tryProcess(
+            () -> Optional.of(config.getNamespaceMapper().getNamespace(sinkRecord, sinkDocument)))
+        .orElse(null);
+  }
+
+  WriteModel<BsonDocument> createWriteModel() {
+    return config.getCdcHandler().isPresent() ? buildWriteModelCDC() : buildWriteModel();
+  }
+
+  private WriteModel<BsonDocument> buildWriteModel() {
+    return tryProcess(
+            () -> {
+              config
+                  .getPostProcessors()
+                  .getPostProcessorList()
+                  .forEach(pp -> pp.process(sinkDocument, sinkRecord));
+              return WriteModelStrategyHelper.createWriteModel(config, sinkDocument);
+            })
+        .orElse(null);
+  }
+
+  private WriteModel<BsonDocument> buildWriteModelCDC() {
+    return tryProcess(
+            () -> config.getCdcHandler().flatMap(cdcHandler -> cdcHandler.handle(sinkDocument)))
+        .orElse(null);
+  }
+
+  private <T> Optional<T> tryProcess(final Supplier<Optional<T>> supplier) {
+    try {
+      return supplier.get();
+    } catch (Exception e) {
+      if (config.logErrors()) {
+        LOGGER.error("Unable to process record {}", sinkRecord, e);
+      }
```

Comment:
I don't believe that is correct: 

> In presence of multiple parameters, if the last argument in a logging statement is an exception, then SLF4J will presume that the user wants the last argument to be treated as an exception instead of a simple parameter

https://www.baeldung.com/slf4j-log-exceptions

Reply:
Thank you! I did not know this about the SLF4J behavior. Found the official docs about this behavior: http://www.slf4j.org/faq.html#paramException